sympy/simplify/cse_main.py

```diff
@@ -536,3 +536,101 @@ def cse(exprs, symbols=None, optimizations=None, postprocess=None,
         return replacements, reduced_exprs
 
     return postprocess(replacements, reduced_exprs)
+
+
+class Marker:
+    pass
+
+# XXX: Should this go in sympy.utilities.iterables?
+def shortest_repeated_subsequence(S, ignore=(Marker,)):
```

Comment:
Seems reasonable. My first goto would have been the sequence matcher from `https://docs.python.org/3/library/difflib.html#difflib.SequenceMatcher` but you want the shortest non-extendable sequence which is not necessarily going to be the longest. 

I was having trouble thinking of an example that needs more than one iteration, but I think something like this is apropos: ABC, ABD, ABD -> `xC, xD, xD -> xC, y, y`.

And I agree that this should go in the iterables file.